Redirect users to https when logging in and accessing the admin

nathanbegbie opened the pull request:
Ticket: https://praekeltorg.atlassian.net/browse/UNICEF-298

This PR will add middleware that will redirect users to use `https` when logging in, registering or accessing the admin.

[Link 1](https://gist.github.com/bryanchow/1035190/c0b060ccee057aa803cfa586d845b2049bb3560a)
[Link 2](https://bitbucket.org/nextscreenlabs/django-ssl-redirect/src/95da4b1eab0ad00a00bcb7b9fb09f02fa8ee7019/ssl_redirect/middleware.py?at=default&fileviewer=file-view-default)

nathanbegbie committed:
Create middleware to redirect users when accessing certain paths

nathanbegbie committed:
Write tests for SSLRedirectMiddleware

nathanbegbie committed:
Add SSLRedirectMiddleware to settings

miltontony commented:
👌 looks good from me. @JayH5 will rely on you for a final 👍 since you discussed the solution with the team.

JayH5 commented on iogt/middleware.py:
```diff
@@ -0,0 +1,27 @@
+
+    def response_should_be_secure(self, request, HTTPS_PATHS):
+        for path in HTTPS_PATHS:
+            if request.path.startswith(u'/{}'.format(path)):
```
Is `request.path` a string (I don't know if there's some Django magic here)? If it is, `startswith` is not going to necessarily match parent paths only, e.g. both `/admin` and `/admin-django` could be matched by `/admin` which may be unexpected. The solution would probably be to parse the path into some kind of proper URL abstraction.

Also, personally I'd just require the `HTTPS_PATHS` setting values to start with `/`--or this would be another thing a URL library would take care of. If somebody _did_ put a path in the settings with a leading `/` it wouldn't work as expected currently.

_Also_, tiny nit... having an uppercase parameter name is weird to me 😅.

JayH5 commented on iogt/tests/test_middleware.py:
```diff
@@ -0,0 +1,57 @@
+
+    def test_no_redirect_when_secure(self):
+        headers = {'HTTP_X_FORWARDED_PROTO': 'https'}
+        request = self.factory.get('/admin/', **headers)
```
Edge case with absent/present trailing `/` should probably be tested.

nathanbegbie replied:
The use of the new `clean_path` ensure that URL paths are represented as a list of segments - so we can rest assured that no matter how it's declared in the settings file, it will be represented consistently in the middleware
e.g. `/admin/login` -> `['admin', 'login']`
`admin/login` -> `['admin', 'login']`
`admin/login/` -> `['admin', 'login']`

JayH5 reviewed (commented):
Sorry for being nitpicky, but if this is a security-critical thing we need to make sure it's fool-proof.

nathanbegbie committed:
rename middleware variable to lower case

nathanbegbie committed:
Create better functions for handling URL format

Make sure that when dealing with URL paths we are using them in a consistent format: a list of strings representing each segment of the path

nathanbegbie committed:
Fix replicated test

JayH5 commented on iogt/middleware.py:
```diff
@@ -0,0 +1,62 @@
+    if path == u'/':
+        return [u'/']
+    else:
+        return filter(lambda a: a != '', path.split('/'))
```
Nothing _technically_ wrong with paths with multiple `/`. We just want to ignore a first leading or final trailing one. But they're also rare/weird enough to maybe not matter.

`filter` will probably make this Python 3-unfriendly.

URLs are hard I'm sorry :(

nathanbegbie replied:
@JayH5 I can update `filter` to `list(filter(...)` to make it python3 compatible.
But do you want me to revert back to checking paths as strings as opposed to lists, assuming I've checked for trailing and leading slashes?

JayH5 replied:
@nathanbegbie list comp is probably best:
```python
return [a for a in path.split(u'/') if a]
```
Still think comparing segments is definitely better than strings.

nathanbegbie committed:
Use list comprehensions instead of filter function

nathanbegbie commented:
@JayH5 ready for re-review

JayH5 reviewed (approved):
:+1:
